add new aws recommendations (#6820)
harness/developer-hub: docs/cloud-cost-management/5-use-ccm-cost-governance/asset-governance/AWS/AWS-recommendations.md

**Permissions Required:** 
- **Dry Run:** 
    - ```elasticache:DescribeCacheClusters```
- **Run Once:** 
    - ```elasticache:DescribeCacheClusters```
    - ```elasticache:DeleteCacheCluster```
    - ```elasticache:DeleteReplicationGroup```

---

### Recommendation: s3-lifecycle-configuration

**Description:** Configure lifecycle for s3 buckets wherever it is absent which would help to reduce storage spend

**Policy Used:**

```yaml
policies:
  - name: s3-lifecycle-configuration
    resource: aws.s3
    description: |
      Configure lifecycle for s3 buckets wherever it is absent which would help to reduce storage spend
    filters:
      - type : value
        key : Lifecycle
        value : absent
    actions:
      - type: configure-lifecycle
        rules:
          - ID: harness-default-lifecycle
            Status: Enabled
            Filter:
              Prefix: ''
            Expiration:
              ExpiredObjectDeleteMarker: True
            AbortIncompleteMultipartUpload:
              DaysAfterInitiation: 7
            NoncurrentVersionExpiration:
              NoncurrentDays: 30
              NewerNoncurrentVersions: 6
```

> **Savings Computed**: 

Reviewer: Missing Savings Computed.

**Permissions Required:**
- **Dry Run:** 
    - ```s3:GetLifecycleConfiguration```
- **Run Once:** 
    - ```s3:GetLifecycleConfiguration```
    - ```s3:PutLifecycleConfiguration```

---

### Recommendation: s3-intelligent-tiering-configuration

**Description:** Configure intelligent tiering for s3 buckets wherever it is disabled which would help to reduce storage spend.

**Policy Used:**

```yaml
policies:
  - name: s3-intelligent-tiering-configuration
    resource: aws.s3
    description: |
      Configure intelligent tiering for s3 buckets wherever it is disabled which would help to reduce storage spend.
    filters:
      - not:
          - type: intelligent-tiering
            attrs:
              - Status: Enabled
    actions:
      - type: set-intelligent-tiering
        Id: harness-default
        IntelligentTieringConfiguration:
          Id: harness-default
          Status: Enabled
          Tierings:
            - Days: 90
              AccessTier: ARCHIVE_ACCESS
            - Days: 180
              AccessTier: DEEP_ARCHIVE_ACCESS
```

> **Savings Computed:** 

Reviewer: Missing Savings Computed.

---

### Recommendation: under-utilized-elasticcache-cluster

> **Description:** Delete underutilised cache cluster with CPU utilisation less than 5% in last 7 days.

Reviewer 2: ```suggestion
**Description:** Delete underutilized cache cluster with CPU utilization less than 5% in the last 7 days.
```

**Policy Used:**

```yaml
policies:
  - name: under-utilized-elasticcache-cluster
    resource: cache-cluster
    description: |
      Delete underutilised cache cluster with CPU utilisation less than 5% in last 7 days
    filters:
      - type: metrics
        name: CPUUtilization
        days: 7
        period: 86400
        value: 5
        op: less-than
    actions:
      - type: delete
        skip-snapshot: false
```

> **Savings Computed:** The policy identifies a list of resources on which potential savings are calculated by summing up cost of each resource for last 30 days.

Reviewer 2: ```suggestion
**Savings Computed:** The policy identifies a list of resources on which potential savings are calculated by summing up the cost of each resource for the last 30 days.
```

**Permissions Required:** 
- **Dry Run:** 
    - ```elasticache:DescribeCacheClusters```
- **Run Once:** 
    - ```elasticache:DescribeCacheClusters```
    - ```elasticache:DeleteCacheCluster```
    - ```elasticache:DeleteReplicationGroup```

> ---

Reviewer 2: ```suggestion
```

### Recommendation: s3-lifecycle-configuration

> **Description:** Configure lifecycle for s3 buckets wherever it is absent which would help to reduce storage spend

Reviewer 2: ```suggestion
**Description:** Configure lifecycle for S3 buckets wherever it is absent which would help to reduce storage spend
```

- Cost Impact: Deleting 40% of noncurrent versions reduces the total storage used by these versions.

**Example Calculation**

Let's assume the following for a single S3 bucket:

**Total Storage Used**: 1 TB (1,024 GB) in the S3 Standard storage class.

**Storage Distribution:**
- Current versions: 50% (512 GB)
- Noncurrent versions: 40% (410 GB)
- Incomplete multipart uploads: 10% (102 GB)

**Calculations:**

<DocImage path={require('../static/policy_comparison.png')} width="100%" height="100%" title="Click to view full size image" />

**Total Savings**
 <DocImage path={require('../static/savings1.png')} width="60%" height="60%" title="Click to view full size image" />

References: 
- [AWS S3 Lifecycle Policies](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lifecycle-mgmt.html)
- [AWS S3 Pricing](https://aws.amazon.com/s3/pricing/)
- [Managing S3 Costs](https://aws.amazon.com/blogs/storage/optimizing-costs-with-amazon-s3-lifecycle-configurations/)

**Permissions Required:**
- **Dry Run:** 
    - ```s3:GetLifecycleConfiguration```
- **Run Once:** 
    - ```s3:GetLifecycleConfiguration```
    - ```s3:PutLifecycleConfiguration```

> ---

Reviewer 2: ```suggestion
```

### Recommendation: s3-intelligent-tiering-configuration

> **Description:** Configure intelligent tiering for s3 buckets wherever it is disabled which would help to reduce storage spend.

Reviewer 2: ```suggestion
**Description:** Configure intelligent tiering for S3 buckets wherever it is disabled which would help to reduce storage spend.
```

              NewerNoncurrentVersions: 6
```

**Savings Computed**: 
To estimate the percentage cost savings from the given S3 lifecycle policies, we need to look at the specific actions and apply some reasonable assumptions. Here's a step-by-step approach:

1. Abort Incomplete Multipart Uploads after 7 days:
- Assumption: 5% of all uploads are incomplete and are not cleaned up without this policy.
- Cost Impact: Each incomplete multipart upload that is aborted saves the storage cost of the data uploaded so far.

2. Expire Noncurrent Versions after 30 days (keeping 6 versions):
- Assumption: Each object has, on average, 10 noncurrent versions stored. Expiring noncurrent versions after 30 days, keeping only the latest 6, will delete 4 out of every 10 noncurrent versions.
- Cost Impact: Deleting 40% of noncurrent versions reduces the total storage used by these versions.

**Example Calculation**

Let's assume the following for a single S3 bucket:

**Total Storage Used**: 1 TB (1,024 GB) in the S3 Standard storage class.

**Storage Distribution:**
- Current versions: 50% (512 GB)
- Noncurrent versions: 40% (410 GB)
- Incomplete multipart uploads: 10% (102 GB)

> **Without Lifecycle Policy**

Reviewer: Can we keep it side by side maybe in a table comparison?

Author: done
